pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -827,6 +835,69 @@ allChainsLabel:
 	return mutable.Listener
 }
 
+func buildSidecarListenerTLSContext(tlsOpts *networking.Server_TLSOptions) *auth.DownstreamTlsContext {
+	// ingress inbound tls only support "SIMPLE" and "MUTUAL" mode
+	if tlsOpts == nil || (tlsOpts.Mode != networking.Server_TLSOptions_SIMPLE && tlsOpts.Mode != networking.Server_TLSOptions_MUTUAL) {
+		return nil
+	}
+
+	tls := &auth.DownstreamTlsContext{
+		CommonTlsContext: &auth.CommonTlsContext{
+			AlpnProtocols: util.ALPNHttp,
+		},
+	}
+
+	tls.CommonTlsContext.TlsCertificates = []*auth.TlsCertificate{
+		{
+			CertificateChain: &core.DataSource{
+				Specifier: &core.DataSource_Filename{
+					Filename: tlsOpts.ServerCertificate,
+				},
+			},
+			PrivateKey: &core.DataSource{
+				Specifier: &core.DataSource_Filename{
+					Filename: tlsOpts.PrivateKey,
+				},
+			},
+		},
+	}
+	var trustedCa *core.DataSource
+	if len(tlsOpts.CaCertificates) != 0 {
+		trustedCa = &core.DataSource{
+			Specifier: &core.DataSource_Filename{
+				Filename: tlsOpts.CaCertificates,
+			},
+		}
+	}
+	if trustedCa != nil || len(tlsOpts.SubjectAltNames) > 0 {
+		tls.CommonTlsContext.ValidationContextType = &auth.CommonTlsContext_ValidationContext{
+			ValidationContext: &auth.CertificateValidationContext{
+				TrustedCa:            trustedCa,
```

Comment:
if trustedCa is nil, this does not cause any issues right?

Reply:
Copied from gateway, will have a look.

Reply:
If `RequireClientCertificate` is false, then `ValidationContextType` will be ignored.
And only in MUTUAL and ISTIO_MUTUAL mod, RequireClientCertificate is set, and in that case, the `trustedCa` will not be nil.  This is very confusing, will refactor it.